### DISCUSSION TITLE
Sharing code between server and app

### DISCUSSION THREAD

**Discussion by fleeser** [serverpod/serverpod#642] [Q&A] [answered]
I am looking for best practices to share code between the server and the Flutter app meaning to make use of the logic in both projects.

When trying to validate inputs from the user for example, it would make sense to check the input before sending a request, but also inside the endpoint for when someone also calls the endpoint from another app or something else. Otherwise in some edge cases it could happen that the developers are not paying attention when updating the logic and do this to only one project.

Therefore I was thinking about if it would be a good approach to put these validator methods inside the _client folder?

**Comment by yahu1031** [comment]
Client code can be modified as per your wish and use it(not recommended). But in the generation again you will lose the whole code you injected. But any kind of validations are better to make in app level instead on the server. If you still would like to do it on the server, then validate before communicating with DB.

**Reply by fleeser** [reply]
I know that it is better to validate inputs before sending them to the server, but I was wondering why the source code then contains something like the following if I could just simply do that before sending the request.

`email = email.trim().toLowerCase();`

**Comment by vlidholt** [comment, marked as answer]
The best practice is to create a new Dart package called `myprojectname_shared`. You put any code you want to share in there. The `myprojectname_client` isn't meant to be used on the server and includes files with conflicting names.
